backend/accounts.go

```diff
@@ -591,25 +591,54 @@ func (backend *Backend) createAndAddAccount(coin coinpkg.Coin, persistedConfig *
 			if err == nil {
 				keystoreName = persistedKeystore.Name
 			}
-			backend.Notify(observable.Event{
-				Subject: "connect-keystore",
-				Action:  action.Replace,
-				Object: data{
-					Type:         "connect",
-					KeystoreName: keystoreName,
-				},
-			})
-			ks, err := backend.connectKeystore.connect(
-				backend.Keystore(),
-				accountRootFingerprint,
-				20*time.Minute,
-			)
+			var ks keystore.Keystore
+			for {
+				backend.Notify(observable.Event{
+					Subject: "connect-keystore",
+					Action:  action.Replace,
+					Object: data{
+						Type:         "connect",
+						KeystoreName: keystoreName,
+					},
+				})
+				ks, err = backend.connectKeystore.connect(
+					backend.Keystore(),
+					accountRootFingerprint,
+					20*time.Minute,
+				)
+				if err == nil || errp.Cause(err) != errWrongKeystore {
+					break
+				} else {
+					backend.Notify(observable.Event{
+						Subject: "connect-keystore",
+						Action:  action.Replace,
+						Object: data{
+							Type:         "error",
+							ErrorCode:    "wrongKeystore",
+							ErrorMessage: err.Error(),
+						},
+					})
+					c := make(chan bool)
+					// retryCallback is called when the current keystore is deregistered or when
+					// CancelConnectKeystore() is called.
+					// In the first case it allows to make a new connection attempt, in the last one
+					// it'll make this function return ErrUserAbort.
+					backend.connectKeystore.retryCallback = func(retry bool) {
```

Comment:
should add a method `setRetryCallback()` which acquires the lock and call that here, otherwise there could be a race condition.

Reply:
Good catch, thanks! Done

Reply:
Thinking about it again, since that code is only executed when a `connect` call returns `wrongKeystore`, and the `connect` call returns something different when a new `connect` is executed, that should be only executed by a single thread at a time. Anyway, using a proper lock is definitely better and safer.